ParProcCo/job_scheduler.py

```diff
@@ -183,163 +374,158 @@ def _create_template(self, i: int) -> drmaa2.JobTemplate:
+        wait_until = time.time() + check_time
+        remaining_jobs = job_ids.copy()
+        terminated_jobs = []
+        while len(remaining_jobs) > 0 and time.time() <= wait_until:
+            for job_id in remaining_jobs:
+                ji, state = self.fetch_and_update_state(job_id)
+                if state in STATEGROUP.ENDED:
+                    remaining_jobs.remove(job_id)
+                    terminated_jobs.append((job_id, ji, state))
+            time.sleep(min(60, check_time))
+
+    def wait_all_jobs_started(self, job_ids: List[int], check_time: int):
+        wait_until = time.time() + check_time
+        remaining_jobs = job_ids.copy()
+        started_jobs = []
+        while len(remaining_jobs) > 0 and time.time() <= wait_until:
+            for job_id in remaining_jobs:
+                _, state = self.fetch_and_update_state(job_id)
+                if state not in STATEGROUP.STARTING:
+                    started_jobs.append(job_id)
+                    remaining_jobs.remove(job_id)
+        return started_jobs
+
+    def _wait_for_jobs(
+        self,
+    ) -> None:
         max_time = int(round(self.timeout.total_seconds()))
         check_time = min(120, max_time)  # 2 minutes or less
         start_wait = max(check_time, 60)  # wait at least 1 minute
         try:
-            job_list = [status_info.job for status_info in self.status_infos]
+            job_list = list(self.status_infos)
             # Wait for jobs to start (timeout shouldn't include queue time)
-            job_list_str = ", ".join([str(job.id) for job in job_list])
-            logging.info(f"Waiting for jobs to start: {job_list_str}")
+            logging.info(f"Waiting for jobs to start: {job_list}")
             jobs_left = len(job_list)
             while jobs_left > 0:
-                jobs_left -= len(session.wait_all_started(job_list, start_wait))
+                jobs_left -= len(self.wait_all_jobs_started(job_list, start_wait))
                 if jobs_left:
                     logging.info(f"Jobs left to start: {jobs_left}")
-            logging.info(f"Jobs started, waiting for jobs: {job_list_str}")
+            logging.info(f"Jobs started, waiting for jobs: {job_list}")
             total_time = 0
             while total_time < max_time and job_list:
-                session.wait_all_terminated(job_list, check_time)
+                self.wait_all_jobs_terminated(job_list, check_time)
                 total_time += check_time
                 jobs_remaining = []
-                for job in job_list:
-                    if job.get_state()[0] == drmaa2.JobState.RUNNING:
-                        jobs_remaining.append(job)
-                        logging.debug(f"Job {job.id} still running")
+                for job_id in job_list:
+                    ji, state = self.fetch_and_update_state(job_id)
+                    if ji.job_state == SLURMSTATE.RUNNING:
+                        jobs_remaining.append(job_id)
```

Comment:
Is this necessary if the `wait_all_jobs_terminated` returned a list of running jobs?